### PR TITLE
chore: store autosynth synth logs as sponge_log.log for use as build artifacts

### PR DESCRIPTION
**Pull request by chingor13**
This allows autosynth to capture individual synthtool runs' output in sponge_log.log files as they are generated. We switch from using a temporary file to using a structured location: `logs/[repo]/[source]/[index]/sponge_log.log` (e.g. `logs/googleapis/java-asset/googleapis/2/sponge_log.log`). These artifacts are uploaded but not yet made public to ResultStore.

### CODE REVIEW

File: autosynth/synthesizer.py [hunk 1]
```diff
@@ -43,10 +45,10 @@ def synthesize(self, environ: typing.Mapping[str, str] = None) -> str:
         pass
 
     def synthesize_and_catch_exception(
-        self, environ: typing.Mapping[str, str] = None
+        self, logfile: pathlib.Path, environ: typing.Mapping[str, str] = None
```
Comment (SurferJeffAtGoogle) [inline autosynth/synthesizer.py:48]: I think the name `log_file_path` would be more consistent with the names in this code base.

File: autosynth/synthesizer.py [hunk 2]
```diff
@@ -96,22 +100,22 @@ def synthesize(self, environ: typing.Mapping[str, str] = None) -> str:
...
+            stdout=tee_proc.stdin,
+            env=(environ or os.environ),
+            universal_newlines=True,
+        )
+        if synth_proc.returncode:
+            logger.error("Synthesis failed")
+            synth_proc.check_returncode()  # Raise an exception.
+
+        with open(logfile, "r") as fp:
+            return fp.read()
```
Comment (SurferJeffAtGoogle) [inline autosynth/synthesizer.py:121]: The synthesize() function returns a string, but line 121 returns a byte array.
Open the file with mode "rt" to read and return a string.

File: autosynth/synth.py [hunk 3]
```diff
@@ -144,6 +144,7 @@ def __init__(
         temp_dir: str,
         metadata_path: str,
         synth_path: str,
+        log_path: pathlib.Path = None,
```
Comment (SurferJeffAtGoogle) [inline autosynth/synth.py:147]: Because it's a path to a directory, let's call it `log_dir_path`.

File: autosynth/synth.py [hunk 4]
```diff
@@ -160,6 +161,10 @@ def __init__(
         self.environ["SYNTHTOOL_PRECONFIG_FILE"] = self._preconfig_path
         self.source_name = ""  # Only non-empty for forks
         self.version_zero = VersionZero()
+        if log_path is None:
```
Comment (SurferJeffAtGoogle) [inline autosynth/synth.py:164]:
```py
self.log_dir_path = log_dir_path or pathlib.Path(tempfile.TemporaryDirectory().name)
```

File: autosynth/synth.py [hunk 5]
```diff
@@ -209,7 +214,7 @@ def fork(self) -> typing.List["SynthesizeLoopToolbox"]:
             [typing.List[SynthesizeLoopToolbox]] -- A toolbox for each source.
         """
         forks = []
-        for i, group in enumerate(self.version_groups):
+        for i, _group in enumerate(self.version_groups):
```
Comment (SurferJeffAtGoogle) [inline autosynth/synth.py:217]: is _group never used?

Comment (chingor13) [reply autosynth/synth.py:217]: Yeah, my IDE suggested the change.

File: autosynth/synth.py (hunk 4 above)
Comment (chingor13) [reply autosynth/synth.py]: Done.

File: autosynth/synth.py (hunk 3 above)
Comment (chingor13) [reply autosynth/synth.py]: Doine

File: autosynth/synthesizer.py (hunk 1 above)
Comment (chingor13) [reply autosynth/synthesizer.py]: Done

File: autosynth/synthesizer.py (hunk 2 above)
Comment (chingor13) [reply autosynth/synthesizer.py:121]: Opened with "rt". Do I need to cast to a string or does `read()` change return type based on the the file mode opened with?

Comment (SurferJeffAtGoogle) [reply autosynth/synthesizer.py:121]: "rt" is sufficient.  read()'s return type varies with the file mode.  :-/

**Pull request merged**